cli/src/commands/gerrit.rs

```diff
@@ -0,0 +1,137 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Debug;
+
+use clap::Subcommand;
+use indexmap::IndexSet;
+use jj_lib::commit::Commit;
+use jj_lib::repo::Repo;
+
+use crate::cli_util::{
+    get_git_repo, resolve_multiple_nonempty_revsets, user_error, CommandError, CommandHelper,
+    RevisionArg,
+};
+use crate::ui::Ui;
+
+/// Interact with Gerrit Code Review.
+#[derive(Subcommand, Clone, Debug)]
+pub enum GerritCommand {
+    /// "Mail in" a revision to Gerrit; this is the primary way to create
+    /// changes or update changes for code review.
+    ///
+    /// This command modifies each commit in the revset to include a `Change-Id`
+    /// footer in its commit message if one does not already exist. This ID is
+    /// NOT compatible with jj IDs, and is a Gerrit-specific ID.
+    ///
+    /// Note: this command takes a revset as an argument, and will mail in all
+    /// revisions in the revset to Gerrit appropriately, so you may post trees
+    /// or ranges of commits to Gerrit for review.
+    ///
+    /// Note: this command does not actually send an email or use a Mail
+    /// Transfer Agent, but rather uploads the revision to Gerrit for review
+    /// through underlying Git logic. The term "mail" is used to disambiguate
+    /// from the word "submit", which is what Gerrit calls "merging an approved
+    /// change into a target branch", and is used to refer to the concept of
+    /// "mailing in" something to be read by someone else.
+    Mail(MailCommand),
+}
+
+pub fn cmd_gerrit(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    subcommand: &GerritCommand,
+) -> Result<(), CommandError> {
+    match subcommand {
+        GerritCommand::Mail(review) => {
+            cmd_mail(ui, command, review)?;
+        }
+    }
+    Ok(())
+}
+
+#[derive(clap::Args, Clone, Debug)]
+pub struct MailCommand {
+    /// The revset, selecting which commits are mailed in to Gerrit. This can be
+    /// any arbitrary set of commits; they will be modified to include a
+    /// `Change-Id` footer if one does not already exist, and then mailed off to
+    /// Gerrit for review.
+    #[arg(long, short = 'r')]
+    revision: Vec<RevisionArg>,
+
+    /// The origin where your changes will be sent. This MUST be a reference of
+    /// the form `branch@remote`, where `branch` is the target branch name that
+    /// the change will be submitted to, and `remote` is the remote to push to.
+    /// This remote MUST be a Git push URL for your Gerrit instance.
+    #[arg(long = "for", short = 'f', default_value = "main@origin")]
```

Comment:
So while `trunk()` probably does make some sense since it tracks what the upstream would be implicitly (without hardcoding pre-existing branch/remote names), I think it's rather confusing because revsets are already used to specify the trees you want to push, and can be specified multiple times, and `trunk()` looks quite like one at a glance. For example,

```
jj gerrit mail -r 'open() ~ empty()' -f 'trunk()'
```

reads quite strangely to me, especially because `trunk()` would be rather special in this context where it actually specifies a pair of `(branch, remote)` to send the change to; it's not clear at a glance what that means. I think that the `branch@remote` syntax is more familiar in this case since it's what shows up in the default log output, it's used for all other `@git` refs, etc. At the same time, it's also reasonably easy to translate to Gerrit's `git push origin HEAD:refs/for/main` so it's not too unfamiliar. It's half-way between us and them.

Related, but what I would actually like to do is have this default be over-rideable in the repository configuration, so that people can also do things like set the URL to the Gerrit instance. e.g.

```
jj config set --repo gerrit.default-remote canon@upstream
```

or something of this manner to override `main@origin` as the default `--for`. We could also have a nested TOML list in the same manner so that remotes can have individually specified default branch targets, e.g. `set gerrit.<remote>.default-for "master"` where `<remote>` can be any Git remote.

Reply:
> jj gerrit mail -r 'open() ~ empty()' -f 'trunk()'

Yeah, it's a bit odd. I don't think `--for` should support a revset in general, but I mean the default could be derived from the `trunk()`. If the user configured per-repository `trunk()`, it would be something like `trunk() = "foo@origin"`, so the purpose is quite similar to `gerrit.default-remote`. Maybe we should have a parsable configuration for that, and both `trunk()` and the gerrit `--for` should default to it?

BTW, is this `origin` in `--for foo@origin` the gerrit endpoint? The project I worked before had a gerrit-specific SSH remote, and we usually don't pull from it. In that setup, the user-visible remote branch is something like `master@origin`, but the remote to push to is `gerrit`.